lib/concurrent/struct/immutable_struct.rb

```diff
@@ -0,0 +1,87 @@
+require 'concurrent/struct/abstract_struct'
+require 'concurrent/synchronization'
+
+module Concurrent
+
+  # A thread-safe, immutable variation of Ruby's standard `Struct`.
+  #
+  # @see http://ruby-doc.org/core-2.2.0/Struct.html Ruby standard library `Struct`
+  module ImmutableStruct
+    include AbstractStruct
+
+    # @!visibility private
+    def initialize(*values)
+      ns_initialize(*values)
+    end
```

Comment:
If you change the way the constructors work in `Synchronization::Object` we'll have to update every class that extends it. I'm fine with that if you think that the call to `#synchronize` in the initializer should be optional. I'm going to merge this PR today since it works with `Synchronization::Object` as it currently exists. Once you finish your updates to the synchronization layer we can go back and update `ImmutableStruct` along with everything else.